### PR TITLE
[skip changelog] Automate license headers checks in CI

### PR DESCRIPTION
**Pull request by alessio-perugini**
## Please check if the PR fulfills these requirements

See [how to contribute](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/)

- [x] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-cli/pulls)
      before creating one)
- [x] The PR follows
      [our contributing guidelines](https://arduino.github.io/arduino-cli/latest/CONTRIBUTING/#pull-requests)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] `UPGRADING.md` has been updated with a migration guide (for breaking changes)
- [ ] `configuration.schema.json` updated if new parameters are added.

## What kind of change does this PR introduce?

This adds a new CI check for new go files missing the license headers. 
We have to put in the root of the folder a template called: `license_header.tpl`. This file will be used by `addlicense` to add any missing license header in the go file.

## What is the current behavior?

If we create a new go file and forget to add a license header we can spot it only by a manual CR which might be error-prone if the PR is large.

## What is the new behavior?

Every time a new golang file is pushed with a missing license header the CI will fail and warn the user.

## Other information

Any suggestion are welcomed

### CODE REVIEW

**Comment by cmaglie** [comment]
I tried to change the license in `main.go`, but running the check would not detect it. I guess it only generates the license header when completely missing?

Can this be changed to detect also if a license is not the correct one?

File: docsgen/main.go [hunk 1]
```diff
@@ -1,8 +1,9 @@
-// Source: https://github.com/arduino/tooling-project-assets/blob/main/workflow-templates/assets/cobra/docsgen/main.go
```
Comment (cmaglie) [inline docsgen/main.go:1]: It's a pity to lose this reference, it should be copied after the license

File: .github/workflows/check-go-task.yml [hunk 2]
```diff
@@ -219,6 +219,55 @@ jobs:
...
+              -type d -name 'testdata' -prune -o \
+              -type d -name 'legacy' -prune -o \
+              -type f ! -name '*_grpc.pb.go' \
+              -type f ! -name 'json_easyjson.go' \
+              -type f ! -name 'index_easyjson.go' \
+              -regex ".*\.go$" -print
+          ); do
+          # We read only the license header, removing all the comments symbols and changing the date to be 2000. This will
+          # help us understand if the other part of the license are exactly the same as the template.
+          HEADER=$(head -n$N_LINES $file | sed -r 's| [0-9]{4} | 2000 |' | sed 's|// ||; s|^//||' )
```
Comment (cmaglie) [inline .github/workflows/check-go-task.yml:263]: We must keep the year range `2020-2023`. I did some research and found that:

1. the starting year must be kept to indicate when the copyright started (otherwise any previous work may not be considered under copyright)
2. the ending year must be kept to indicate when the copyright expires (it happens after the last year + some decades actually...)
3. we cannot use `2020-present`...
4. old headers should be updated every time a file is touched (for example if a file made in 2020 is modified in 2023 then the ending year in the header must be updated with `2020-2023`)

At this point, maybe the simplest solution is to just put the year range in the template file and update the template  + all source files every year?

File: .github/workflows/check-go-task.yml [hunk 3]
```diff
@@ -219,6 +219,55 @@ jobs:
...
+          if [ "$LICENSE" != "$HEADER" ]; then
+            STATUS=1
+            echo "$file"
+          fi
+          done
+          exit $STATUS
```
Comment (cmaglie) [inline .github/workflows/check-go-task.yml:269]: I don't like this check because it just tells you that a file is wrong but didn't tell you what's wrong, and we don't have any way to re-run it locally. Also running `task go:add-license-headers` won't fix the wrong file.

Ideally, the action should just run `task go:add-license-header` and check if there are changes with `git diff`.
(more or less the same as the check for protoc files https://github.com/arduino/arduino-cli/blob/master/.github/workflows/check-protobuf-task.yml#L130).

In other words `task go:add-license-header` should write the files as they should be before pushing the PR.

**Review by cmaglie — CHANGES_REQUESTED** [review]

**Comment by alessio-perugini** [comment]
Ideally, we'd like to have a tool that can recognize and fix unwanted changes in the license header without being only limited to checking the presence of a license header. We tried to implement something quick with bash, but it's more tricky than expected. For example, do we need to provide a date range for our license headers? Or it's okay to keep the year when the file was created? In case we have to opt for a year range, should we always put 2020-CURRENT_YEAR?

We agreed that the most common mistake is to forget to add the license for a newly created file. Without overcomplicating this, the `addlicense` should be enough for now.

**Review by cmaglie — APPROVED** [review]
